### DISCUSSION TITLE
Header/Query/Path parameters with same name fills them all

### DISCUSSION THREAD

**Discussion by zoliknemet** [fastapi/fastapi#8398] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

import uvicorn
from fastapi import FastAPI, APIRouter, Depends, Path, Header, Query
from pydantic import PositiveInt

app = FastAPI()
router = APIRouter()


def get_user_id_from_header(user_id: Union[PositiveInt, None] = Header(default=None)):
    return user_id


def get_user_id_from_query(user_id: Union[PositiveInt, None] = Query(default=None)):
    return user_id


@router.get("/user/{user_id}")
async def article(
    user_id: PositiveInt = Path(),
    header_user_id=Depends(get_user_id_from_header),
    query_user_id=Depends(get_user_id_from_query),
):
    return {
        "user_id": user_id,
        "header_user_id": header_user_id,
        "query_user_id": query_user_id,
    }


app.include_router(router)

if __name__ == "__main__":
    uvicorn.run(app)
```

### Description

Request
```
curl -X 'GET' \
  'http://localhost:8000/user/123' \
  -H 'accept: application/json'
```
Response:
```
{
  "user_id": 123,
  "header_user_id": 123,
  "query_user_id": 123
}
```
Response is suppose to be this:
```
{
  "user_id": 123,
  "header_user_id": null,
  "query_user_id": null
}
```

I know this is not so good by design, but sometimes we want to add some global dependencies and we must be sure that the names of the header/path/query are unique on each endpoint on whole app.

I presume there is already issue like this, but I can not find it.

### Operating System

Linux

### FastAPI Version

0.68.2 and 0.85.1

### Python Version

3.7.14 and 3.10.6

**Comment by JarroVGIT** [comment]
Although I hadn't expected that behaviour, it does make (some sort of) sense. Using the same parameter names across dependencies apparently has some unforeseen side effects. I haven't pinned down the troubling code, but I do have a workaround for your:

```python
def get_user_id_from_header(user_id_header: Union[PositiveInt, None] = Header(default=None, alias="user_id")):
    return user_id_header


def get_user_id_from_query(user_id_query: Union[PositiveInt, None] = Query(default=None, alias="user_id")):
    return user_id_query
```

This ensures you can still use `user_id` as either query or header parameter (or both) without these side effects:
```
curl -X 'GET' \
  'http://localhost:8000/user/123' \
  -H 'accept: application/json'

RESPONSE:
{
  "user_id": 123,
  "header_user_id": null,
  "query_user_id": null
}
```
and
```
curl -X 'GET' \
  'http://localhost:8000/user/123?user_id=456' \
  -H 'accept: application/json' \
  -H 'user_id: 789'

RESPONSE:
{
  "user_id": 123,
  "header_user_id": 789,
  "query_user_id": 456
}
```

**Comment by zoliknemet** [comment]
I am aware of workaround.
I still think this should be fixed, it is not natural.
If FastAPI wants to be serious framework it is time to get rid of that kind of design failures and errors...

Code where this happens is in `solve_dependencies` method and it is not easy to fix it, but this part of code needs serious refactoring. It is too complex to be a recursion, it takes a lot of input arguments, and it returns too many values, simply said it breaks every software design principle.

Problematic part is here:
```python
    path_values, path_errors = request_params_to_args(
        dependant.path_params, request.path_params
    )
    query_values, query_errors = request_params_to_args(
        dependant.query_params, request.query_params
    )
    header_values, header_errors = request_params_to_args(
        dependant.header_params, request.headers
    )
    cookie_values, cookie_errors = request_params_to_args(
        dependant.cookie_params, request.cookies
    )
    values.update(path_values)
    values.update(query_values)
    values.update(header_values)
    values.update(cookie_values)

```
when everything is put in `values` we lose information from where it came from
later this code is run
```python
solved = await run_in_threadpool(call, **sub_values)
```
where `call` is dependency method that user wrote - in this example: `get_user_id_from_header` and `sub_values` is from above `values` dictionary - it consists of key/values like this `{"user_id": 123}` but it took it from `path_values` and not headers.

I would like to help, and I understand most of the codebase of FastaAPI, and I would refactor this part entirely, but I know that pull request will never be merged or even considered...


EDIT: @JarroVGIT thank you for your comment!

**Comment by YuriiMotov** [comment]
Code fails with `AssertionError: Cannot use `Header` for path param 'user_id` in current version
